### DISCUSSION TITLE
Solving arrays + quantified bitvectors

### DISCUSSION THREAD

**Discussion by TDacik** [cvc5/cvc5#9324] [Q&A] [answered]
I tried cvc5 on several formulae containing arrays and quantification over bitvectors. On those formulae, cvc5 always immediately returns _unknown_ with reason _incomplete_. Are there any options that can force cvc5 to try to solve such formulae? I tried several options from the Quantifier and Bitvector modules, but none worked. 

One example is the following formula which is solved immediately by z3 and bitwuzla:
```smt
(set-logic ABV)

(declare-const footprint0 (_ BitVec 3))
(declare-const heap (Array (_ BitVec 3) (_ BitVec 3)))
(declare-const nil (_ BitVec 3))

(assert
  (and
    (bvule nil (_ bv2 3))
    (bvule (select heap (_ bv3 3)) (_ bv2 3))
    (bvule (select heap (_ bv2 3)) (_ bv2 3))
    (bvule (select heap (_ bv1 3)) (_ bv2 3))
    (bvule (select heap (_ bv0 3)) (_ bv2 3))
    (not (distinct (bvand (bvshl (_ bv1 3) nil) footprint0) (_ bv0 3)))
    (= nil (select heap nil))
    (not
      (or
        (exists ((FP2!0 (_ BitVec 3)) (FP1!0 (_ BitVec 3)))
          (and
            (= footprint0 (bvor (bvor (bvor (_ bv0 3) (_ bv0 3)) FP1!0) FP2!0))
            (= (bvand FP1!0 FP2!0) (_ bv0 3)) (not (= FP2!0 (_ bv0 3)))
            (not (= FP1!0 (_ bv0 3)))
          )
        )
        (= footprint0 (_ bv0 3))
      )
    )
  )
)
(check-sat)
```

**Comment by ajreynol** [comment, marked as answer]
Thanks for the benchmark, there are two options that solve this:
`--cegqi-full`, which enables a more complete version of the techniques from  https://homepage.divms.uiowa.edu/~ajreynol/cav18.pdf
`--sygus-inst`, see https://homepage.divms.uiowa.edu/~ajreynol/tacas21.pdf

Unfortunately neither of these are enabled by default.  It may be possible for us to revisit if the former can be enabled by default.
